File: scripts/fix_license_headers.py
#!/usr/bin/env python3
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Normalise Apache 2.0 license headers across the repo.

Targets ``.py``, ``.sh``, ``.yml``, ``.yaml``, ``Dockerfile``. Two fixes:

1.  Replace ``Copyright 2026 Xerxes-Agents Author`` with
    ``Copyright 2026 Xerxes-Agents Author`` (drops the "The" article).
2.  When a file has the opening license lines but is missing the trailing
    "Unless required by applicable law…" block, splice the block in
    immediately after the ``LICENSE-2.0`` line.

Idempotent — running it a second time is a no-op.
"""

from __future__ import annotations

import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

WRONG_COPYRIGHT = "Copyright 2026 Xerxes-Agents Author"
RIGHT_COPYRIGHT = "Copyright 2026 Xerxes-Agents Author"

LICENSE_URL_LINE = "#     https://www.apache.org/licenses/LICENSE-2.0"

TRAILER = (
    "#\n"
    "# Unless required by applicable law or agreed to in writing, software\n"
    '# distributed under the License is distributed on an "AS IS" BASIS,\n'
    "# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n"
    "# See the License for the specific language governing permissions and\n"
    "# limitations under the License."
)

TRAILER_MARKER = "Unless required by applicable law"

EXTS = {".py", ".sh", ".yml", ".yaml"}
FILE_NAMES = {"Dockerfile"}

SKIP_DIRS = {".git", ".venv", "node_modules", "__pycache__", "build", "dist", ".ruff_cache", ".pytest_cache"}


def should_visit(path: Path) -> bool:
    parts = set(path.parts)
    if parts & SKIP_DIRS:
        return False
    if path.suffix in EXTS:
        return True
    if path.name in FILE_NAMES:
        return True
    return False


def fix_text(text: str) -> tuple[str, list[str]]:
    """Return (new_text, list_of_fixes_applied)."""
    fixes: list[str] = []
    new_text = text

    if WRONG_COPYRIGHT in new_text:
        new_text = new_text.replace(WRONG_COPYRIGHT, RIGHT_COPYRIGHT)
        fixes.append("the-article")

    if LICENSE_URL_LINE in new_text and TRAILER_MARKER not in new_text:
        # Splice the trailer in directly after the LICENSE_URL_LINE. We do this
        # only once per file (first occurrence), which is what we want for
        # any sane header.
        idx = new_text.index(LICENSE_URL_LINE)
        end_of_line = new_text.index("\n", idx)
        new_text = new_text[: end_of_line + 1] + TRAILER + "\n" + new_text[end_of_line + 1 :]
        fixes.append("trailer")

    return new_text, fixes


def main() -> int:
    changed: list[tuple[Path, list[str]]] = []
    for path in REPO_ROOT.rglob("*"):
        if not path.is_file() or not should_visit(path):
            continue
        try:
            text = path.read_text(encoding="utf-8")
        except (UnicodeDecodeError, OSError):
            continue
        new_text, fixes = fix_text(text)
        if not fixes or new_text == text:
            continue
        path.write_text(new_text, encoding="utf-8")
        changed.append((path, fixes))

    print(f"Files changed: {len(changed)}")
    for path, fixes in sorted(changed, key=lambda p: str(p[0])):
        rel = path.relative_to(REPO_ROOT)
        print(f"  {rel}  [{', '.join(fixes)}]")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/typescript/src/maintenance/fixLicenseHeaders.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { readdir } from 'node:fs/promises'
import { extname, join, relative, resolve } from 'node:path'

import { errorMessage, readUtf8Text, writeTextFile, type WriteOptions } from './io.js'

export const CANONICAL_COPYRIGHT_LINE = '# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).'
export const LICENSE_URL_LINE = '#     https://www.apache.org/licenses/LICENSE-2.0'
export const LICENSE_TRAILER_MARKER = 'Unless required by applicable law'
export const LICENSE_TRAILER = [
  '#',
  '# Unless required by applicable law or agreed to in writing, software',
  '# distributed under the License is distributed on an "AS IS" BASIS,',
  '# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.',
  '# See the License for the specific language governing permissions and',
  '# limitations under the License.',
].join('\n')

const COPYRIGHT = /^# Copyright 2026 (?:The )?Xerxes-Agents Author(?: @erfanzar \(Erfan Zare Chavoshi\))?\.?$/gmu
const EXTENSIONS = new Set(['.py', '.sh', '.yml', '.yaml'])
const FILE_NAMES = new Set(['Dockerfile'])
const SKIPPED_DIRECTORIES = new Set([
  '.git',
  '.pytest_cache',
  '.ruff_cache',
  '.venv',
  '__pycache__',
  'build',
  'dist',
  'node_modules',
])

export interface LicenseFix {
  fixes: readonly ('copyright' | 'trailer')[]
  path: string
}

export interface LicenseHeaderRun extends LicenseFix {
  changed: boolean
}

/** Normalize Xerxes Apache headers in source files below one explicit root directory. */
export async function fixLicenseHeaders(root: string, options: WriteOptions = {}): Promise<LicenseHeaderRun[]> {
  const normalizedRoot = resolve(root)
  const changes: LicenseHeaderRun[] = []

  for (const path of await targetFiles(normalizedRoot)) {
    let source: string
    try {
      source = await readUtf8Text(path)
    } catch {
      continue
    }
    const fixed = fixLicenseText(source)
    if (fixed.fixes.length === 0 || fixed.text === source) continue
    await writeTextFile(path, fixed.text, options)
    changes.push({ changed: options.dryRun !== true, fixes: fixed.fixes, path })
  }

  return changes.sort((left, right) => left.path.localeCompare(right.path))
}

/** Return the normalized text and the exact repairs that were necessary. */
export function fixLicenseText(source: string): { fixes: readonly ('copyright' | 'trailer')[]; text: string } {
  const fixes: ('copyright' | 'trailer')[] = []
  let text = source.replace(COPYRIGHT, CANONICAL_COPYRIGHT_LINE)
  if (text !== source) fixes.push('copyright')

  if (text.includes(LICENSE_URL_LINE) && !text.includes(LICENSE_TRAILER_MARKER)) {
    const urlOffset = text.indexOf(LICENSE_URL_LINE)
    const newline = text.indexOf('\n', urlOffset)
    const insertion = newline === -1 ? text.length : newline + 1
    const separator = newline === -1 ? '\n' : ''
    text = `${text.slice(0, insertion)}${separator}${LICENSE_TRAILER}\n${text.slice(insertion)}`
    fixes.push('trailer')
  }

  return { fixes, text }
}

export async function main(args: readonly string[] = process.argv.slice(2)): Promise<number> {
  try {
    const parsed = parseArguments(args)
    if (parsed === undefined) return 0
    const changes = await fixLicenseHeaders(parsed.root, { atomic: parsed.atomic, dryRun: parsed.dryRun })
    console.log(`${parsed.dryRun ? 'Files that would change' : 'Files changed'}: ${changes.length}`)
    for (const change of changes) {
      const path = relative(parsed.root, change.path) || change.path
      console.log(`  ${path}  [${change.fixes.join(', ')}]`)
    }
    return 0
  } catch (error) {
    console.error(`fix-license-headers: ${errorMessage(error)}`)
    return 1
  }
}

if (import.meta.main) {
  process.exitCode = await main()
}

async function targetFiles(root: string): Promise<string[]> {
  const files: string[] = []
  const directories = [root]

  while (directories.length > 0) {
    const directory = directories.pop()
    if (directory === undefined) continue
    const entries = await readdir(directory, { withFileTypes: true })
    for (const entry of entries.sort((left, right) => left.name.localeCompare(right.name))) {
      const path = join(directory, entry.name)
      if (entry.isDirectory()) {
        if (!SKIPPED_DIRECTORIES.has(entry.name)) directories.push(path)
        continue
      }
      if (entry.isFile() && shouldVisit(entry.name)) files.push(path)
    }
  }

  return files
}

function parseArguments(args: readonly string[]): { atomic: boolean; dryRun: boolean; root: string } | undefined {
  let atomic = true
  let dryRun = false
  let root = resolve(process.cwd())

  for (let index = 0; index < args.length; index += 1) {
    const argument = args[index]
    if (argument === undefined) continue
    if (argument === '--help' || argument === '-h') {
      console.log('Usage: bun fixLicenseHeaders.ts [--root <directory>] [--dry-run] [--no-atomic]')
      return undefined
    }
    if (argument === '--dry-run') {
      dryRun = true
      continue
    }
    if (argument === '--no-atomic') {
      atomic = false
      continue
    }
    if (argument === '--root') {
      const candidate = args[index + 1]
      if (candidate === undefined) throw new Error('--root requires a directory path.')
      root = resolve(candidate)
      index += 1
      continue
    }
    throw new Error(`Unknown option: ${argument}`)
  }

  return { atomic, dryRun, root }
}

function shouldVisit(fileName: string): boolean {
  return FILE_NAMES.has(fileName) || EXTENSIONS.has(extname(fileName))
}
